frontend/src/features/public-form/PublicFormProvider.tsx

```diff
@@ -47,20 +47,64 @@ interface PublicFormProviderProps {
   children: React.ReactNode
 }
 
+export function useCommonFormProvider(formId: string) {
+  const [vfnTransaction, setVfnTransaction] =
+    useState<FetchNewTransactionResponse>()
+  const miniHeaderRef = useRef<HTMLDivElement>(null)
+  const { createTransactionMutation } = useTransactionMutations(formId)
+  const toast = useToast({ isClosable: true })
+  const vfnToastIdRef = useRef<string | number>()
+  const desyncToastIdRef = useRef<string | number>()
+
+  const getTransactionId = useCallback(async () => {
+    if (!vfnTransaction || isPast(vfnTransaction.expireAt)) {
+      const result = await createTransactionMutation.mutateAsync()
+      setVfnTransaction(result)
+      return result.transactionId
+    }
+    return vfnTransaction.transactionId
+  }, [createTransactionMutation, vfnTransaction])
+
+  const isNotFormId = useMemo(() => !FORMID_REGEX.test(formId), [formId])
+
+  const expiryInMs = useMemo(() => {
+    if (!vfnTransaction?.expireAt) return null
+    return differenceInMilliseconds(vfnTransaction.expireAt, Date.now())
+  }, [vfnTransaction])
```

Comment:
also can probably be different for public/preview. might be able to be done in another PR

Reply:
refer to above comment related to `getTransactionId `